cmd/cmd_linux.go

```diff
@@ -115,37 +115,38 @@ func coreSupportsReExec(corePath string) bool {
 }
 
 // InternalToolPath returns the path of an internal snapd tool. The tool
-// *must* be located inside /usr/lib/snapd/.
+// *must* be located inside the same tree as the current binary.
 //
 // The return value is either the path of the tool in the current distribution
-// or in the core snap (or the ubuntu-core snap). This handles spiritual
-// "re-exec" where we run the tool from the core snap if the environment allows
-// us to do so.
-func InternalToolPath(tool string) string {
+// or in the core/snapd snap (or the ubuntu-core snap) if the current binary is
+// ran from that location.
+func InternalToolPath(tool string) (string, error) {
 	distroTool := filepath.Join(dirs.DistroLibExecDir, tool)
 
 	// find the internal path relative to the running snapd, this
 	// ensure we don't rely on the state of the system (like
 	// having a valid "current" symlink).
 	exe, err := osReadlink("/proc/self/exe")
 	if err != nil {
-		logger.Noticef("cannot read /proc/self/exe: %v, using tool outside core", err)
-		return distroTool
+		return "", err
 	}
 
-	// ensure we never use this helper from anything but
-	if !strings.HasSuffix(exe, "/snapd") && !strings.HasSuffix(exe, ".test") {
-		log.Panicf("InternalToolPath can only be used from snapd, got: %s", exe)
-	}
+	if !strings.HasPrefix(exe, dirs.DistroLibExecDir) {
```

Comment:
What will happen if I run `snapd` inside my git tree for testing? Will it no longer find the helpers in /usr/lib/snapd ?

Reply:
easiest is to write a test for the  /home/dev/snapd/snapd case ?

Reply:
We do some magic around `/usr/`, so `/home/dev/snapd/snapd` will work. Added a test nonetheless.